test/infrastructure/docker/e2e/docker_suite_test.go

```diff
@@ -113,6 +113,30 @@ var _ = BeforeSuite(func() {
 })
 
 var _ = AfterSuite(func() {
+	deleteClusterInput := framework.DeleteClusterInput{
+		Deleter: mgmtClient,
+		Cluster: cluster,
+	}
+	framework.DeleteCluster(ctx, deleteClusterInput)
+
+	waitForClusterDeletedInput := framework.WaitForClusterDeletedInput{
+		Getter:  mgmtClient,
+		Cluster: cluster,
+	}
+	framework.WaitForClusterDeleted(ctx, waitForClusterDeletedInput)
+
+	assertAllClusterAPIResourcesAreGoneInput := framework.AssertAllClusterAPIResourcesAreGoneInput{
+		Lister:  mgmtClient,
+		Cluster: cluster,
+	}
+	framework.AssertAllClusterAPIResourcesAreGone(ctx, assertAllClusterAPIResourcesAreGoneInput)
+
+	ensureDockerDeletedInput := ensureDockerArtifactsDeletedInput{
+		Lister:  mgmtClient,
+		Cluster: cluster,
+	}
+	ensureDockerArtifactsDeleted(ensureDockerDeletedInput)
```

Comment:
Do these helpers fail (somewhat) gracefully when `mgmtClient` or `cluster` are nil? If not, then this might fail during teardown prior to getting the logs below or tearing down the Management cluster.

Though it looks like we'll already fail before tearing down the management cluster if we fail to get the logs

Reply:
You are right, they do fail. Trying to find a way to make it continue on failure.

Reply:
Maybe deletion shouldn't be in AfterSuite and should instead be its own set of assertions that run after the tests run.

It'd be cool if we could do something like FOCUS='Create|Upgrade|ScaleDown|ScaleUp|Delete'. Basically be able to control exactly which operations are running. Unfortunately ginkgo doesn't give us a great way to do that except try and abuse FOCUS, but that is surely fraught with peril.

Reply:
@chuckha the problem here is that you should not assume ordering between the various specs that match focus.

In general specs that are defined withing the same `Describe` are run in the order they are defined (unless running in parallel), but it does not hold up across different `Describe` blocks.

Reply:
Yeah, this is the general assumption I broke when originally writing these tests, mostly for the sake of test-running-speed. I really don't want to spin up a new cluster for each test, but working with ginkgo instead of against it will likely yield us better results code wise and worse results speed wise.

Perhaps one option (getting way out of scope of the PR now...) is to make the before each is as quick as possible by spinning up a single node control plane cluster with no workers instead of a 3 node control plane. That would put the runtime somewhere around 4 minutes for each test's start up time plus the time the test takes to run (an additional ~2 minutes per control plane that gets created). But it would make the code organization much better and allow for parallelization.

The other idea I had was to manage the ordering of tests internally, pass some kind of flag to the test run that lists the tests you want to run  (-tests="upgrade,scaleup,scaledown") and then the tests keep track of the order and run the tests in the correct order skipping tests that are not selected. However, this ruins any possibility of parallelization.